### PR TITLE
New remote tools - first part

### PR DESCRIPTION
**Pull request by sergiocazzolato**
These tools are the base to run nested tests, but also can be used to run checks on remote instances. 

The remote.setup is a tool that is used to configure how to connect to the remote instance. The idea is to run in prepare-each this tool so then the connection is done transparently.

remote.exec is used to run commands in the remote instance (nested.tests exec)
remote.push is used to send local files to the remote instance (nested.tests copy)
remote.pull is used to retrieve remote files to the local instance (nested.tests get)

All the tools are tested with their own test.

### CODE REVIEW

File: remote/remote.exec [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+show_help() {
+    echo "usage: remote.exec [--user <USER>] [--pass <PASS>] <CMD>"
+    echo ""
+    echo "Available options:"
+    echo "  -h --help   show this help message."
+    echo ""
+}
+
+_load_config() {
```
Comment (Meulengracht) [inline remote/remote.exec:11]: is it just for coding style we prefix it to indicate internal function or is there also some kind of bash magic to this?

File: remote/remote.exec [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    CFG_FILE="$(remote.setup get-config-path)"
+    if [ ! -f "$CFG_FILE" ]; then
+        echo "remote.exec: config file \"$CFG_FILE\" not found, please run remote.setup command first"
+        return 0
```
Comment (Meulengracht) [inline remote/remote.exec:16]: sorry for a stupid question, but it looks to me we just supress this error message, do we handle it at all? it looks like we just try to connect with empty user/pass?

File: remote/remote.pull [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+show_help() {
+    echo "usage: remote.pull <REMOTE_PATH> <LOCAL_PATH>"
+    echo ""
+    echo "Available options:"
+    echo "  -h --help   show this help message."
+    echo ""
+}
+
+_load_config() {
```
Comment (Meulengracht) [inline remote/remote.pull:11]: same comments as above in this file

File: remote/remote.pull [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+_get_cert() {
+    if [ -n "$TESTS_REMOTE_PASS" ]; then
+        return
+    elif [ -n "$TESTS_REMOTE_CERT" ]; then
+        echo "-i $TESTS_REMOTE_CERT"
+    fi
+}
+
+remote_pull() {
+    REMOTE_PATH=$1
```
Comment (Meulengracht) [inline remote/remote.pull]: should these be local?

File: remote/remote.push [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    echo ""
+    echo "Available options:"
+    echo "  -h --help   show this help message."
+    echo ""
+}
+
+_load_config() {
+    CFG_FILE="$(remote.setup get-config-path)"
+    if [ ! -f "$CFG_FILE" ]; then
+        echo "remote.push: config file \"$CFG_FILE\" not found, please run remote.setup command first"
```
Comment (Meulengracht) [inline remote/remote.push:15]: same here as the others, I guess it will be for all of them

File: remote/remote.push [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+_get_cert() {
+    if [ -n "$TESTS_REMOTE_PASS" ]; then
+        return
+    elif [ -n "$TESTS_REMOTE_CERT" ]; then
+        echo "-i $TESTS_REMOTE_CERT"
+    fi
+}
+
+remote_push() {
+    LOCAL_PATH=$1
```
Comment (Meulengracht) [inline remote/remote.push]: also local?

File: remote/remote.wait-for [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+exec() {
+    SSH_PASS="$(_get_pass)"
+    SSH_CERT="$(_get_cert)"
+    "$SSH_PASS" ssh "$SSH_CERT" -p "$TESTS_REMOTE_PORT" -o ConnectTimeout=10 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no "$TESTS_REMOTE_USER"@"$TESTS_REMOTE_HOST" "$@"
+}
+
+push() {
+    LOCAL_PATH=$1
```
Comment (Meulengracht) [inline remote/remote.wait-for]: local here?

File: remote/remote.wait-for [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+        *)
+            echo "tests.remote: unsupported parameter $1" >&2
+            exit 1
+            ;;
+    esac
+
+    "$action" "$@"
+}
+
+prepare() {
```
Comment (Meulengracht) [inline remote/remote.wait-for]: empty function?

**Review by Meulengracht — APPROVED** [review]
Just some minor comments otherwise it looks really good!

File: remote/remote.exec (hunk 1 above)
Comment (sergiocazzolato) [reply remote/remote.exec:11]: hehehe, it is not black magic, as you pointed it is to indicate it is an internal function

File: remote/remote.exec (hunk 2 above)
Comment (sergiocazzolato) [reply remote/remote.exec:16]: @Meulengracht well, not sure if we should try with empty user, but we could have a default one? either ubuntu or root perhaps? empty password is ok

File: remote/remote.pull (hunk 4 above)
Comment (sergiocazzolato) [reply remote/remote.pull]: yes, I'll update all the files using local vars

File: remote/remote.push (hunk 6 above)
Comment (sergiocazzolato) [reply remote/remote.push]: done

File: remote/remote.wait-for (hunk 7 above)
Comment (sergiocazzolato) [reply remote/remote.wait-for]: done

File: remote/remote.wait-for (hunk 8 above)
Comment (sergiocazzolato) [reply remote/remote.wait-for]: mmm, wait for is going in the second part of the change (next pr). Ill remove it

Comment (sergiocazzolato) [reply remote/remote.wait-for]: thanks for the review

File: remote/remote.exec (hunk 2 above)
Comment (sergiocazzolato) [reply remote/remote.exec:16]: I was thinking about it and it is pretty complex to define default values.
I'll try to add that to the following PR, I think we need to discuss it a bit more.

**Pull request merged**